src/python/pants/backend/jvm/subsystems/zinc.py

```diff
@@ -192,9 +196,36 @@ def zinc(self):
 
   @property
   def dist(self):
-    """Return the distribution selected for Zinc.
+    """Return the `Distribution` selected for Zinc based on execution strategy.
 
-    :rtype: list of str
+    :rtype: pants.java.distribution.distribution.Distribution
+    """
+    underlying_dist = self.underlying_dist
+    if self._execution_strategy != NailgunTaskBase.HERMETIC:
+      # symlink .pants.d/.jdk -> /some/java/home/
+      jdk_home_symlink = os.path.relpath(
+        os.path.join(self._zinc_factory.get_options().pants_workdir, '.jdk'),
+        get_buildroot())
+
+      # Since this code can be run in multi-threading mode due to multiple
+      # zinc workers, we need to make sure the file operations below is atomic.
+      with Lock():
```

Comment:
I _think_ the way locks work in Python, you need to allocate this to a variable, and have someone call `acquire()` and `release()` on it, but I've never actually used locks in Python

Reply:
the with statement is a thing. https://docs.python.org/2.7/library/threading.html#using-locks-conditions-and-semaphores-in-the-with-statement

Reply:
I get the `with` part, but I think each call to this function will have its own `Lock` instance, so they won't mutually exclude each other, right?

Reply:
hmm makes sense. I can't find any doc saying it's global, so it's safe to assume it's not, so I will extract to a class variable.

Reply:
Yep, needs to be a class variable.

@illicitonion : Are usecases like this sufficient to justify adding `snapshot2 = snapshot1.rename(srcpath, destpath)` at this point (likely factored as a static method, I guess)? I don't think this is the first time we've had to contort to make the "global file namespace" thing work.

Reply:
This is all for the non-hermetic case, where snapshots aren't involved.